text/0000-if-let-chains.md

```diff
@@ -0,0 +1,2003 @@
+We can then debate alternatives, in particular using `EXPR match PAT`,
+more rigorously post shipping Rust 2018. Finalizing the syntax and can
+then be decided in a tracking issue or another RFC.
+
+## Irrefutable let bindings after the first refutable binding
+
+Should temporary and irrefutable `let`s without patterns be allowed as
+in the following example?
+
+```rust
+if let &List(_, ref list) = meta
+    && let mut iter = list.iter().filter_map(extract_word) // <-- Irrefutable
+    && let Some(ident) = iter.next()
+    && let None = iter.next() {
+    *set = Some(syn::Ty::Path(None, ident.clone().into()));
+} else {
+    error::param_malformed();
+}
+```
+  
+With normal `if let` expressions, this is an error as seen with the 
+following example:
+
+```rust
+fn main() {
+    if let x = 1 { 2 } else { 3 };
+}
+```
+
+Compiling the above ill-formed program results in:
+
+```
+error[E0162]: irrefutable if-let pattern
+```
+
+[RFC 2086]: https://github.com/rust-lang/rfcs/pull/2086
+
+However, with the implementation of [RFC 2086], this error will instead
+become a warning.  This is understandable - while the program could have
+perfectly well defined semantics, where the value of the expression is
+always 2, allowing the form would invite some developers to write in a
+non-obvious way. A warning is however a good middle ground.
+
+However, when let bindings in the middle are irrefutable, there is some value
+in not warning against the construct. In the case of the initial example in this
+subsection, it would be written as follows without irrefutable let bindings:
+
+```rust
+if let &List(_, ref list) = meta {
+   let mut iter = list.iter().filter_map(extract_word);
+    if let Some(ident) = iter.next() &&
+       let None = iter.next() {
+        *set = Some(syn::Ty::Path(None, ident.clone().into()));
+    } else {
+        error::param_malformed();
+    }
+} else {
+    error::param_malformed();
+}
+```
+
+However, now we have introduced rightward drift and duplication again,
+which we wanted to avoid.
+
+On the other hand, allowing irrefutable patterns in the middle without a
+warning may give the impression that the irrefutable pattern is refutable,
+or cast doubt on it making semantics possibly harder to grasp quickly.
+
+This is a tricky question, which we leave open for consideration during
+the stabilization period or even after stabilization.
+
+## Chained `if let`s inside `match` arms
+
+Would the following be accepted by a Rust compiler?
+
+```rust
+match EXPR {
+    PAT if let PAT = EXPR && EXPR && ... => { .. }
+    _ => { .. }
+}
+```
+
+[RFC 2294]: https://github.com/rust-lang/rfcs/pull/2294
+
+The combination of the accepted, but yet to be stabilized, [RFC 2294], and this
+RFC would entail that it would be accepted. However, at this point, and in the
+interest of time, we leave this for a future RFC or for pre-stabilization.
+
+# Appendix A - Style considerations
+
+How should the features introduced in this RFC be formatted?
```

Comment:
6.

```rust
if let foo = bar &&
if let x = y &&
z <= 42
{
}
```

No indent at all, just a list of conditions

Reply:
Wouldn't this be the following instead?

```rust
if let foo = bar &&
if let x = y &&
if z <= 42 {
    ...
}
```

Reply:
Well since I agree now that the `if`-spammy syntax is not good:

```rust
if let foo = bar &&
let x = y &&
z <= 42 {
    ...
}
```

Reply:
Updated the style discussion with this version now :)